[Core] Introduce Process interface

Kunchd opened the pull request:
## Description
This PR introduces an interface to abstract away `Process`. This is done as part of the effort to better define how each component in ray should interact with each other. Additionally, the interface also introduces the ability for us to define `fake_process` dummy classes for testing instead of attempting to construct dummy `Process` from the actual process class. This will provide the ability to perform unit tests that isolates and restrict the tested errors to a smaller area.

## Related issues
Other PRs that introduces interfaces for better modularity and testing: https://github.com/ray-project/ray/pull/57177

## Additional information
N/A

<!-- BUGBOT_STATUS --><sup><a href="https://cursor.com/dashboard?tab=bugbot">Cursor Bugbot</a> found 1 potential issue for commit <u>ca110cc</u></sup><!-- /BUGBOT_STATUS -->

Commit:
Introduce process interface

Signed-off-by: davik <davik@anyscale.com>

dayshah reviewed (commented):
Do we really need the factory here seems like it should just be a function passed into classes that returns a Process, and in tests it can return the fake process. Maybe two functions because you need the pid construction and args construction.

And the worker pool should take these functions in its constructor in main.cc instead of every CreateWorker call taking a process, so that you can test all of the worker pool in the future + test from a higher level

Commit:
Remove process factory to construct processes directly using ctor

This commit removes the process factory as it did not provide any
additional logic outside of wrapping the constructors. Additionally
This commit also address comment style nits.

Signed-off-by: davik <davik@anyscale.com>

Kunchd commented on src/ray/util/process_factory.cc:
```diff
@@ -0,0 +1,108 @@
+
+namespace ray {
+
+std::unique_ptr<ProcessInterface> ProcessFactory::CreateNull() {
```
However, that change requires changing how many of the worker_pool functions returns the error code (currently done through output parameter). I believe this is outside the scope of this PR.

dayshah replied:
ya whether you want to use StatusSetOr or optional or nullptr is up for debate, but it's all better than constructing an empty Process

Kunchd commented on src/ray/util/process_factory.h:
```diff
@@ -0,0 +1,109 @@
+   * @return Any error that occurred.
+   */
+  static std::error_code Call(const std::vector<std::string> &args,
+                              const ProcessEnvironment &env = {});
```
From offline discussions, we've concluded that factories were not necessary as they do not provide any meaningful functionality other than wrapping the constructors. These functions will remain in `process.h`.

Kunchd commented on src/ray/util/process_factory.cc:
```diff
@@ -0,0 +1,108 @@
+                                   new_process_group);
+}
+
+std::pair<std::unique_ptr<ProcessInterface>, std::error_code> ProcessFactory::Spawn(
```
The error code comes from `fork` and `execvpe`. I'll convert them to `Status` and return `StatusOr`.

Commit:
Add ray check to ensure new worker process is actually added

Signed-off-by: davik <davik@anyscale.com>

Commit:
Merge branch 'master' into process_interface

Kunchd commented on src/ray/raylet/worker_pool.h:
```diff
@@ -930,6 +933,9 @@ class WorkerPool : public WorkerPoolInterface {
   WorkerPoolMetrics worker_pool_metrics_;
 
+  /// A static null process used for returning references in error cases.
+  static const std::unique_ptr<ProcessInterface> kNullProcess;
```
Done!

Kunchd commented on src/ray/util/process.cc:
```diff
@@ -405,18 +377,9 @@ intptr_t ProcessFD::GetFD() const { return fd_; }
-Process::~Process() {}
-
-Process::Process() {}
+Process::~Process() = default;
```
Removed

israbbani commented on src/ray/common/tests/BUILD.bazel:
```diff
@@ -119,6 +119,7 @@ ray_cc_test(
     deps = [
         "//src/ray/common:id",
         "//src/ray/common:memory_monitor",
+        "//src/ray/util:process",
```
Its odd that the tests are written against the concretion and not the interface. I can see creating an instance for the tests, but I would expect the interface to be imported and used for the actual interactions.

Kunchd replied:
The test imports process to use the static `Exec` function to get the stats of free memory. Since the test doesn't actually interact with the process class, it doesn't make too much sense to include the process interface. 

Another way to clean this up might be to directly implement the functionality to get free bytes in memory monitor.

israbbani commented on src/ray/common/memory_monitor.cc:
You didn't add this, but can you delete this comment.

Kunchd commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -163,15 +164,16 @@ WorkerPool::WorkerPool(instrumented_io_context &io_service,
 WorkerPool::~WorkerPool() {
-  std::unordered_set<Process> procs_to_kill;
-  for (const auto &entry : states_by_lang_) {
+  absl::flat_hash_map<pid_t, std::unique_ptr<ProcessInterface>> procs_to_kill;
```
Changed data structure to be keyed by pid to maintain the same behavior as the original implementation. This is because Process is hashed by its pid.

israbbani commented on src/ray/util/process_fd.h:
```diff
@@ -0,0 +1,55 @@
+
+class ProcessFD {
+  pid_t pid_;
+  intptr_t fd_;
```
Explicitly label these are private instead of relying on the default C++ behavior.

israbbani commented on src/ray/util/process_fd.h:
```diff
@@ -0,0 +1,55 @@
+  ProcessFD(const ProcessFD &other) = delete;
+  ProcessFD &operator=(const ProcessFD &other) = delete;
+
+  void CloseFD();
```
This should be private. It's only called in the destructor. Could probably just inline this and call `close` directly in the destructor.

israbbani commented on src/ray/util/process_interface.h:
```diff
@@ -0,0 +1,90 @@
+  virtual bool IsValid() const = 0;
+
+  /// Forcefully kills the process.
+  /// Unsafe for unowned processes.
```
What does this mean?

israbbani commented on src/ray/util/process_utils.h:
```diff
@@ -0,0 +1,67 @@
+#include <system_error>
+#include <vector>
+
+#include "ray/util/compat.h"
```
I think we're importing `compat.h` in a few places to rely on transitive includes. Can we remove compat and include the header files explicitly?

Kunchd replied:
As far as I could tell, compat was included to use the `pid_t` type. So it seems fine?

israbbani commented on src/ray/util/process.h:
Is the `io_service` used anywhere?

israbbani commented on src/ray/util/fake_process.h:
```diff
@@ -0,0 +1,102 @@
+    auto proc = std::make_shared<FakeProcess>(pid);
+    proc->SetAlive(true);
+    return proc;
+  }
```
Is this used anywhere?

Kunchd replied:
Removed

israbbani commented on src/ray/util/fake_process.h:
```diff
@@ -0,0 +1,102 @@
+    return exit_code_;
+  }
+
+  // Test control methods
```
Can you explain a little bit more what this means? I think you're saying these are all methods exclusive to the Fake that are not part of ProcessInterface that can be used to manipulate the internal state of the Fake for testing.

israbbani reviewed (commented):
Thank you for doing this. This is going to help improve a critical and complicated piece of our codebase. I'll probably need to do at least one more round of review on this. Here are a few high-level comments

- Read all the code that’s undocumented in this PR at least once and then improve the documentation at least a little bit.
   - Convert the docstrings to Doxygen. We haven’t turned it on yet for the codebase, but newer code should favour it. Look at the cgroup code for examples.
- Process and ProcessFD can probably be merged into one class.
- Remove transitive includes through compat.h.
- As much as is reasonable, prune out dead code. If the code is only called through the `cpp` codebase, we should eventually just move it there.   

Also good to track a few follow ups. Make an issue and link it to the relevant code
- Remove the dependency of `cpp` on the process targets. It’s not exposed to the other language front-ends and should also not be for cpp.
- Separate x-platform code into its own .cc implementations and put it behind a single interface.

Kunchd commented on src/ray/common/memory_monitor.h:
```diff
@@ -18,7 +18,7 @@
 #include "ray/common/asio/instrumented_io_context.h"
 #include "ray/common/asio/periodical_runner.h"
-#include "ray/util/process.h"
+#include "ray/util/compat.h"
```
Compat is needed to make sure pid_t is defined correctly according to the OS. I've brought it back to `Process.h` so we can maintain the existing include logic instead.

Kunchd replied:
Memory monitor was previously getting the `pid_t` through transitive include.

Kunchd commented on src/ray/util/fake_process.h:
```diff
@@ -0,0 +1,102 @@
+    auto proc = std::make_shared<FakeProcess>();
+    proc->SetNull(true);
+    return proc;
+  }
```
Removed. I will change fake process with pid to be a live by default.

Kunchd commented on src/ray/util/process.h:
```diff
@@ -109,69 +65,62 @@ class Process {
+  /// \param[in] args The command-line arguments.
+  /// \param[in] env Additional environment variables.
+  /// \return Any error that occurred.
   static std::error_code Call(const std::vector<std::string> &args,
```
The static functions that creates/spawns processes are mainly used as convenience functions during testing.

Commit:
Merge Process_FD with Process

Signed-off-by: davik <davik@anyscale.com>

Kunchd commented on src/ray/util/process_fd.h:
```diff
@@ -0,0 +1,55 @@
+
+class ProcessFD {
+  pid_t pid_;
+  intptr_t fd_;
```
The file descriptor is the read end to the pipe with the child where the child has the write end. I've updated the function documentation to describe this.

Commit:
Address comments for logging, documentation, and removing old code

Signed-off-by: davik <davik@anyscale.com>

Commit:
Merge branch 'master' into process_interface

Commit:
Debugging worker pool test IP

Signed-off-by: davik <davik@anyscale.com>

Commit:
Merge branch 'master' into process_interface

Commit:
Debug worker pool tests to support process pointers

Signed-off-by: davik <davik@anyscale.com>

Commit:
Merge branch 'master' into process_interface

Commit:
Add more ray check logging on why error occured in process helper

Signed-off-by: davik <davik@anyscale.com>

israbbani commented on src/ray/util/process.cc:
```diff
@@ -359,121 +103,264 @@ ProcessFD::ProcessFD(pid_t pid, intptr_t fd) : pid_(pid), fd_(fd) {
+  return *this;
+}
+
+StatusOr<std::pair<pid_t, intptr_t>> Process::Spawnvpe(
```
Yeesh this thing is ugly. I'm assuming you kept the behavior the same as before.

israbbani commented on src/ray/util/BUILD.bazel:
```diff
@@ -100,6 +100,29 @@ ray_cc_library(
     ],
 )
 
+ray_cc_library(
```
visibility for bazel targets plz

israbbani commented on src/ray/raylet/worker_pool.h:
```diff
@@ -816,14 +820,15 @@ class WorkerPool : public WorkerPoolInterface {
-                        const rpc::RuntimeEnvInfo &runtime_env_info,
-                        const std::vector<std::string> &dynamic_options,
-                        std::optional<absl::Duration> worker_startup_keep_alive_duration);
+  const std::unique_ptr<ProcessInterface> &AddWorkerProcess(
```
Can we return a `const ProcessInterface&` instead?

israbbani commented on src/ray/common/status.h:
```diff
@@ -446,6 +447,8 @@ class RAY_EXPORT Status {
     return Status(StatusCode::PermissionDenied, msg);
   }
 
+  static Status FromError(const std::error_code &error);
```
plz write a docstring

Kunchd replied:
Concluded that the status refactor should be in a separate PR.

Kunchd replied:
Since I decided that the status refactor should be in a separate PR. This will be removed from this PR.

israbbani reviewed (commented):
It's looking pretty close to done. A couple of comments:

1. Whenever you're returning a Status/StatusOr, can you document which statuses get returned in which cases in the docstring? 
2. The `const std::unique_ptr<T> &` is sketchy. You probably want `const &`. I would even limit that and put an API that lets you get what you want?

Kunchd committed:
Merge branch 'master' into process_interface

Kunchd commented on src/ray/util/process_interface.h:
```diff
@@ -0,0 +1,81 @@
+
+  /**
+   * @brief Waits for process to terminate.
+   * @details Not supported for unowned processes.
```
More details added

Kunchd commented on src/ray/util/process.h:
```diff
@@ -14,162 +14,170 @@
+   *          Behaviors of duplicate environment variables will be undefined.
+   * @return A pair of the process ID and the fd of the read end
+   *         of the pipe to track child process's lifetime on success.
+   *         Returns an error status wrapping the error code on failure.
```
Upon digging into how the status code is ultimately propagated back up as an error code, I've actually decided to not include returning a status as part of this PR. 

The main reason for this is that the current overwrite of error code with status only partially overwrites the code path. At some point, these status are converted back into error code again. However, the conversion from status back to system error code actually return the status code (defined in `status.h`) instead of the actual system error code. Since this actually causes an information loss, I've decided that the status refactor needs to be in a separate PR that completes the refactor for the entire code path.

Commit:
Refactor to use const ref instead of const ref pointer

Signed-off-by: davik <davik@anyscale.com>

Commit:
Merge branch 'master' into process_interface

israbbani commented on src/ray/util/process_interface.h:
```diff
@@ -0,0 +1,83 @@
+
+/**
+ * @class ProcessInterface
+ * @details This interface is used to abstract the process implementation
```
I would limit these comments to what the class is doing. Encapsulation and dependency injection are pretty well understood and don't need to be documented.

israbbani reviewed (commented):
LGTM. Check to see if the process object's lifecycle is handled properly for all cases so we don't get segfaults on un-itialized process pointers. 

- Driver
- IO Workers
- Workers
- Agents

Ping to 🚢

Commit:
Address comments for process interface

Signed-off-by: davik <davik@anyscale.com>

Kunchd committed:
Merge branch 'master' into process_interface

Commit:
Address uninitialized status in wait

Signed-off-by: davik <davik@anyscale.com>

Kunchd committed:
Merge branch 'master' into process_interface

Kunchd committed:
Merge branch 'master' into process_interface

israbbani reviewed (approved):
🚢 when CI passes.

Kunchd committed:
Merge branch 'master' into process_interface

Commit:
Fix copy compiler complaint in worker pool test

Signed-off-by: davik <davik@anyscale.com>

Commit:
Merge branch 'master' into process_interface

Commit:
Fix uninitialized last_started_worker_process

Signed-off-by: davik <davik@anyscale.com>

Kunchd committed:
Merge branch 'master' into process_interface

Commit:
Fix typo in process.cc for windows

Signed-off-by: davik <davik@anyscale.com>

Commit:
Merge branch 'master' into process_interface

Commit:
Enforce Process either construct both pid and fd or none invariant

Signed-off-by: davik <davik@anyscale.com>

Kunchd committed:
Merge branch 'master' into process_interface

Kunchd committed:
Merge branch 'master' into process_interface

israbbani commented on src/ray/raylet/worker.cc:
```diff
@@ -15,6 +15,7 @@
 #include "ray/raylet/worker.h"
 
 #include <boost/bind/bind.hpp>
+#include <iostream>
```
Looks real, but lets merge this PR first.

israbbani reviewed (approved):
There's on must-do follow-up which to fix the RAY_CHECK in Worker::SetProcess. Lets do it separately because this PR is huge. 

I triggered the cpp ubsan/asan/kitchen-sink tests. 

Ping for merge when everything passes.

Kunchd committed:
Merge branch 'master' into process_interface

Commit:
Define execvpe for apple

Signed-off-by: davik <davik@anyscale.com>

Commit:
Merge branch 'process_interface' of github.com:Kunchd/ray into process_interface

Kunchd committed:
Merge branch 'master' into process_interface
